kong/tools/stream_api.lua

```diff
@@ -0,0 +1,111 @@
+
+local stream_api = {}
+
+local _endpoints = {}
+
+
+function stream_api.register_endpoint(path, handler)
+  _endpoints[path] = handler
+end
+
+
+local stream_mt = {}
+stream_mt.__index = stream_mt
+
+function stream_api.serve()
+  local req = setmetatable({ socket = ngx.req.socket() }, stream_mt)
+
+  local path = req:get_path()
+  local handler = path and _endpoints[path]
+  if handler then
+    req:exit(handler(req))
+
+  else
+    req:exit(req:response("404 Not Found"))
+  end
+end
+
+
+function stream_mt:get_line()
+  if not self.req_line then
+    self.req_line = self.socket:receive("*l")
+  end
+
+  return self.req_line
+end
+
+function stream_mt:get_path()
+  if not self.path then
+    self.path = self:get_line():match("^%S+%s+(%S+)")
+  end
+
+  return self.path
+end
+
+function stream_mt:get_headers()
+  if not self.headers then
+    local headers = {}
+
+    while true do
+      local l = self.socket:receive("*l")
+      if l == "" then
+        self.headers = headers
```

Comment:
This line seems unnecessary, as we're doing `self.headers = headers` after the loop.

Reply:
done